src/components/pages/userSubmissionsListPage.tsx

```diff
@@ -1,7 +1,135 @@
+
+type OrderBy = 'submittedAt' | 'dueAt' | 'assignmentName' | 'courseName'
+type GroupBy = 'id' | 'assignmentId' | 'courseId'
+
+const orderByOptions: Option<OrderBy>[] = [
+  { label: 'Submitted At', value: 'submittedAt' },
+  { label: 'Due At', value: 'dueAt' },
+  { label: 'Assignment', value: 'assignmentName' },
+  { label: 'Course', value: 'courseName' },
+]
+
+const groupByOptions: Option<GroupBy>[] = [
+  { label: 'None', value: 'id' },
+  { label: 'Assignment', value: 'assignmentId' },
+  { label: 'Course', value: 'courseId' },
+]
+
+const UserCoursesListPage = () => {
+  const defaultOrderBy = LocalStorageService.get<OrderBy>(ORDER_BY_STORAGE_KEY) || 'submittedAt'
+  const defaultGroupBy = LocalStorageService.get<GroupBy>(GROUP_BY_STORAGE_KEY) || 'id'
+
+  const [loading, setLoading] = useState(true)
+  const [error, setError] = useState(null)
+  const [submissions, setSubmissions] = useState(new Array<Submission>())
+  const [courses, setCourses] = useState<Record<string, Course>>({})
+  const [assignments, setAssignments] = useState<Record<string, Assignment>>({})
+
+  const [orderBy, setOrderBy] = useState<OrderBy>(defaultOrderBy)
+  const [groupBy, setGroupBy] = useState<GroupBy>(defaultGroupBy)
+
+  useEffect(() => {
+    fetchData(orderBy, groupBy)
+  }, [])
+
+  const fetchData = async (orderBy: OrderBy, groupBy: GroupBy) => {
+    try {
+      const submissions = await RequestService.get<Submission[]>(
+        `/api/submissions?orderBy=${orderBy}?groupBy=${groupBy}`,
+      )
+
+      const courseRequests = submissions.map((s) => RequestService.get<Course>(`/api/courses/${s.courseId}`))
+      const assignmentRequests = submissions.map((s) =>
+        RequestService.get<Assignment>(`/api/assignments/${s.assignmentId}`),
+      )
+
+      const courses = await Promise.all(courseRequests)
+      const assignments = await Promise.all(assignmentRequests)
+
+      // Mapify course & assignment ids so we can look them up more easilly with our submission
+      const courseMap: Record<string, Course> = {}
+      for (const course of courses) courseMap[course.id || ''] = course
+
+      const assignmentMap: Record<string, Assignment> = {}
+      for (const assignment of assignments) assignmentMap[assignment.id || ''] = assignment
+
+      setSubmissions(submissions)
+      setAssignments(assignmentMap)
+      setCourses(courseMap)
+    } catch (error) {
+      setError(error)
+    } finally {
+      setLoading(false)
+    }
+  }
+
+  const handleFilterChange = (updatedOrderBy: OrderBy) => {
+    setOrderBy(updatedOrderBy)
+    fetchData(updatedOrderBy, groupBy)
+
+    LocalStorageService.set(ORDER_BY_STORAGE_KEY, updatedOrderBy)
+  }
+
+  const handleGroupByChange = (updatedGroupBy: GroupBy) => {
+    setGroupBy(updatedGroupBy)
+    fetchData(orderBy, updatedGroupBy)
+
+    LocalStorageService.set(GROUP_BY_STORAGE_KEY, updatedGroupBy)
+  }
+
+  if (loading) return <LoadingOverlay delay={250} />
```

Comment:
Just out of curiosity, would giving the LoadingOverlay be best practice for all pages that use it, or is that more dependent of the page itself?

Reply:
That's a reasonable take I think. The main reason I haven't is because I think it'd be counter intuitive to what I'd think a component would do. Like when I go to use something like a loading overlay I wouldn't ever really expect it to delay to render, that just sounds super not intuitive ya know?

Beyond that there are instances where you wouldn't want this on a delay. I think this delay will only ever really be used for page loading.

Reply:
I'm gonna keep it for now cause of the counterintuativeness that I'd mentioned, but I do think that it's reasonable to consider defaulting the delay to some non zero value.

If others disagree, then we can change it over.

Reply:
That sounds pretty reasonable. Given that cases exist where we wouldn't want a delay, and the expected component behavior you mentioned, it makes more sense to have including that delay be something the developer has to actively enable.